File: 98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    bool res = true;
public:
    bool isValidBST(TreeNode* root) {
        vector <int> nums;
        inorder(root, nums);
        return res;
    }
    
    void inorder(TreeNode* root, vector <int>& nums)
    {
        if (root == NULL)
            return;
        if (res)
        {
            inorder(root->left, nums);
            if (nums.size() == 0 || (nums.size() > 0 && root->val > nums[nums.size() - 1]))
                nums.push_back(root->val);
            else
                res = false;
            inorder(root->right, nums);
        }
    }
};

File: 98. Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        self.res = True
        self.inorder(root, [])
        return self.res
    
    def inorder(self, root, nums):
        if not root:
            return
        if self.res:
            self.inorder(root.left, nums)
            if not nums or (nums and root.val > nums[len(nums)-1]):
                nums += [root.val]
            else:
                self.res = False
            self.inorder(root.right, nums)
